File: solutions/707/main.cpp
#include <gtest/gtest.h> 

using namespace std;

class MyLinkedList {
public:
    /** Initialize your data structure here. */
    struct ListNode {
        int val; 
        ListNode* next; 
        ListNode(int x) : val(x), next(nullptr) {}
    };

    MyLinkedList() {
        _dummyHead = new ListNode(0); 
        _size = 0; 
    }
    
    /** Get the value of the index-th node in the linked list. If the index is invalid, return -1. */
    int get(int index) {
        if (index > (_size - 1) || index < 0) {
            return -1; 
        }
        ListNode* cur = _dummyHead->next; 
        while (index--) {
            cur = cur->next; 
        }
        return cur->val; 
    }
    
    /** Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list. */
    void addAtHead(int val) {
        ListNode* newNode = new ListNode(val); 
        newNode->next = _dummyHead->next; 
        _dummyHead->next = newNode; 
        _size++; 
    }
    
    /** Append a node of value val to the last element of the linked list. */
    void addAtTail(int val) {
        ListNode* cur = _dummyHead; 
        while (cur->next != nullptr) {
            cur = cur->next; 
        }
        ListNode* newNode = new ListNode(val);
        cur->next = newNode; 
        _size++; 
    }
    
    /** Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted. */
    void addAtIndex(int index, int val) {
        if (index > _size)
            return; 
        ListNode* newNode = new ListNode(val);
        ListNode* cur = _dummyHead; 
        while (index--) {
            cur = cur->next; 
        }
        newNode->next = cur->next; 
        cur->next = newNode; 
        _size++; 
    }
    
    /** Delete the index-th node in the linked list, if the index is valid. */
    void deleteAtIndex(int index) {
        if (index < 0 || index > _size - 1)
            return; 
        ListNode* cur = _dummyHead; 
        while (index--) {
            cur = cur->next; 
        }
        ListNode* tmp = cur->next; 
        cur->next = tmp->next; 
        delete tmp; 
        _size--; 
    }
private: 
    ListNode* _dummyHead; 
    int _size; 
};

TEST(Test707, SimpleTest)
{
    int result;

    MyLinkedList myLinkedList = MyLinkedList();
    myLinkedList.addAtHead(1);
    myLinkedList.addAtTail(3);
    myLinkedList.addAtIndex(1, 2);    // linked list becomes 1->2->3
    result = myLinkedList.get(1);              // return 2
    EXPECT_EQ(result, 2); 
    myLinkedList.deleteAtIndex(1);    // now the linked list is 1->3
    result = myLinkedList.get(1);              // return 3
    EXPECT_EQ(result, 3); 
}

int main() {
    testing::InitGoogleTest(); 
    return RUN_ALL_TESTS(); 
}

File: solutions/707/main.py
class Node:
    def __init__(self, val, pre, nex): 
        self.val = val 
        self.pre = pre 
        self.nex = nex 

class MyLinkedList:

    def __init__(self):
        self.head = Node(-1, None, None)
        self.tail = Node(-1, None, None)
        self.head.nex = self.tail 
        self.tail.pre = self.head 
        self.len = 0 

    def get(self, index: int) -> int:
        if index < 0 or index >= self.len: 
            return -1 
        node = self.head 
        while index >= 0: 
            node = node.nex 
            index -= 1 
        return node.val 

    def addAtHead(self, val: int) -> None:
        node = Node(val, self.head, self.head.nex)
        self.head.nex.pre = node 
        self.head.nex = node 
        self.len += 1 

    def addAtTail(self, val: int) -> None:
        node = Node(val, self.tail.pre, self.tail)
        self.tail.pre.nex = node 
        self.tail.pre = node 
        self.len += 1

    def addAtIndex(self, index: int, val: int) -> None:
        if index > self.len:
            return 
        self.len += 1
        node = self.head 
        while index > 0: 
            node = node.nex 
            index -= 1 
        new_node = Node(val, node, node.nex)
        node.nex.pre = new_node
        node.nex = new_node 

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.len: 
            return 
        self.len -= 1
        node = self.head 
        while index >= 0: 
            node = node.nex 
            index -= 1 
        node.pre.nex = node.nex 
        node.nex.pre = node.pre 